docs/ilab-model-engine.md

```diff
@@ -0,0 +1,204 @@
+# Redesign ilab to be a Full Fledged Model Engine
+
+This documents describes a new structure for ilab, consisting of sub-commands that act as parent or "group" commands. Currently the only group command is `ilab`. If the models `ilab` produces are meant to be the golden standard for open source AI, the tool to use and manage these models needs to be full fledged and extensible, especially to models we have not produced but should be able to run.
+
+open source container tools like podman an docker are commonly viewed as engines. `ilab` should be a model engine, managing the deployment and creation of AI models. To be clear, I will be using tools like podman as a structural analogy here with full understanding they have different implications and use-cases.
+
+
+## Key Component
+
+### InstructLab Structure Redesign
+
+Here are two outlines. One that represents the state of `ilab` after this enhancement and the second describes `ilab` after I introduce new key commands to this structure. I am including the second outline for clarity purposes, but this EP only covers the contents of the first structure.
+
+```console
+ilab
+|
+|_______model
+|       |
+|       |____convert
+|       |____download
+|       |____train (--convert)
+|       |____serve   (-i)
+|       |____chat
+|       |____inference 
+|       |
+|_______data
+|       |
+|       |____generate
+|       |
+|_______config
+|       |
+|       |____init
+|       |
+|_______tax
+|       |
+|       |____diff
+|       |____check
+|       |____download
+```
+
+vs. currently:
+
+```console
+ilab
+|
+|_____chat
+|_____check
+|_____convert
+|_____download
+|_____generate
+|_____init
+|_____list
+|_____serve
+|_____test
+|_____train
+```
+
+The main point of this structure is to introduce a group hierarchy and to remove some confusions of the existing `ilab` structure. Some key things missing in `ilab` currently are:
+
+1. `ilab init` does too many things.
+    - We don't want "black box" commands. This creates bad UX and also ties too many processes together in which one might eventually need to be changed.
+2. ambiguity in "what" we are generating, training etc.
+    - The current structure of `ilab` requires an in depth knowledge of the ordering of commands and background knowledge for how our CLI is built. This is not the best way to grow. Commands that are not intuitive and require a deep dive into the docs just to get started might not engage users as we hope to. having an intuitive structure in which people can see easily from the docstring what each commands does and how they might use them, makes more sense. We are generating data, then training a model, running a server to interact with that model, then finally chatting with the model
+3. some commands like `ilab test` and `ilab list` should be renamed to show their true purpose (`ilab model inference`and `ilab taxonomy diff`)
```

Comment:
@khaledsulayman @alimaredia WDYT?

Reply:
what about `evaluate`? I know this is a very minimal test compared to our full-scale model evaluation but I think it would help to a) differentiate from code testing and b) mirror some of the other subcommands which mostly seem to be more minimal versions of the larger operations we run on the backend

Reply:
additionally, I do think, minimal as it currently is, evaluation is _really_ what the current `ilab test` is actually doing, and maybe by being more explicit about the correlation between this and the backend version, we can start to think about what is minimal enough within the backend to incorporate into the CLI

Reply:
the only issue with `evaluate` is we might overload that later on with actual eval if we expose from the cli

Reply:
Discussed with Charlie in person but just to update @hickeyma I think that's the point. The CLI is really meant to be a laptop version of the backend pipeline we are building, so having the flow of the CLI mirror it is a plus.

If we feel like the functionality currently handled by `ilab test` is still too minimal to be called evaluation, even for running on a laptop, then we can scope out future work to beef it up as desired.

Reply:
switched to evaluate